### PR TITLE
Separate bindings to different crate.

### PR DESCRIPTION
**Pull request by mephisto41**
This patch do following things.
1. Move all binding stuff to separate crate. So the binding stuff is not related to webrender anymore. Hence, if we want to update webrender version, we can replace the webrender folder directly. This makes update webrender more easily.
2. Add bindings.cpp. In the past, gecko won't re-link webrender lib if we modify webrender. We have to touch any file under gfx to force re-link webrender lib. I add this dummy file to tell gecko re-link the lib when webrender is modified. 
3. Add build script. Now we can type `python build_gecko_webrender.py -r` in the root directory to trigger webrender build instead of change directory to gfx/webrender.

@jrmuizel , Do you think those changes make sense?

### CODE REVIEW

**Comment by kvark** [comment]
Sounds good! I wonder if we can do `./mach build_wr` instead of calling python directly.

**Comment by mstange** [comment]
```
13:28 <@ted> mstange: so, we have pretty good support for rust crates in the build system now
13:29 <@ted> put the crate in the tree somewhere
13:29 <@ted> list it under dependencies here: https://dxr.mozilla.org/mozilla-central/source/toolkit/library/rust/shared/Cargo.toml
13:29 <@ted> then do `mach vendor rust` to get all the crates it depends on vendored into third_party/rust
13:29 <@ted> and it'll get linked into libxul
```

**Comment by staktrace** [comment]
FYI I tried doing the `mach vendor rust` thing and didn't get very far on OS X. First I ran into mach not recognizing git worktrees, so I fixed that in bug 1313446. Then `mach vendor rust` runs `cargo install cargo-vendor` which has all sorts of other dependencies - you need to have cmake and OpenSSL at least, probably more things. I stopped after discovering those since I don't want to go down the road of installing the world right now.

**Comment by jrmuizel** [comment]
1. I'd eventually like to get the bindings part of the upstream but in the mean time I agree it probably makes sense for it to be in a separate crate.
2. I feel like there should be a better solution than this, but maybe it's best just to hack it for now to at least help people avoid mistakes.
3. We should look toward having a better solution than this too.

Other comments:
- can you avoid doing the removal of the webrender_traits prefix in the same commit?
- What's the reason for moving things around?

**Comment by mephisto41** [comment]
> Sounds good! I wonder if we can do ./mach build_wr instead of calling python directly.

As Markus said, gecko can build rust directly. Once it works, we should able to use mach to build webrender.

> FYI I tried doing the mach vendor rust thing and didn't get very far on OS X. First I ran into mach not recognizing git worktrees, so I fixed that in bug 1313446. Then mach vendor rust runs cargo install cargo-vendor which has all sorts of other dependencies - you need to have cmake and OpenSSL at least, probably more things. I stopped after discovering those since I don't want to go down the road of installing the world right now.

Great, I'll try it next.

> ```
> We should look toward having a better solution than this too.
> ```

I agreed. In the future I'll try to use rust in gecko to build webrender which is Markus suggest.

> ```
> What's the reason for moving things around?
> ```

I just think it would be more organized. But I think original is good as well. I'll revert it in next patch

**Comment by mephisto41** [comment]
I have built with `mach vendor rust` and `mach build` with webrender successfully. Will update patch soon.

**Comment by staktrace** [comment]
I also managed to get a successful build with `mach vendor rust` and `mach build` after some finagling. That was on Linux, though, and the build crashes on startup because it can't initialize a GL context. I think that's unrelated. However, when I applied those patches on OS X the build failed because of a missing freetype library. I filed [this issue](https://github.com/servo/rust-freetype/issues/42) for it. My patches so far are sitting at https://github.com/staktrace/gecko-dev/commits/wr-vendor - the 0c2a307 patch is the interesting one.

**Comment by staktrace** [comment]
freetype has been published to crates.io, so my patches (I just updated them) at https://github.com/staktrace/gecko-dev/commits/wr-vendor should work now. I tested it on OS X and @changm will test it on Windows. Note that you need to add `ac_add_options --enable-rust` to your mozconfig to ensure the build system builds libgkrust and webrender as part of it.

**Comment by staktrace** [comment]
Also we'll probably want to make webrender a "feature" (in the [Cargo.toml sense](http://doc.crates.io/manifest.html#the-features-section)) of libgkrust that's disabled by default, so that when we merge to m-c we don't suddenly bloat libxul. We can do that in a separate set of patches afterwards.

**Comment by mephisto41** [comment]
Great! Since you integrated webrender into libgkrust in issue 31. This pr will focus on separate bindings to different crate and make webrender a feature of libgkrust.

**Comment by mephisto41** [comment]
I have added --enable-webrender option so that you can include or exclude webrender by this option. Currently pr depend on kats's patches.

**Comment by staktrace** [comment]
Nice, thanks! I think we can probably start landing some of your work (like the --enable-webrender) option and the ifdefs even without the rest of my vendoring patches. I have some other things I can extract from my WIPs so maybe I'll make a new PR with that stuff to merge it sooner rather than later.

File: gfx/layers/opengl/GLManager.cpp [hunk 1]
```diff
@@ -91,10 +94,14 @@ GLManager::CreateGLManager(LayerManagerComposite* aManager)
 /* static */ GLManager*
 GLManager::CreateGLManager(WebRenderLayerManager* aManager)
```
Comment (staktrace) [inline gfx/layers/opengl/GLManager.cpp]: This function signature uses WebRenderLayerManager\* outside of the ifdef. Doesn't that fail compilation if webrender is turned off?

Comment (mephisto41) [reply gfx/layers/opengl/GLManager.cpp]: Yap, since it only use pointer type. Forward decalaration makes compiler happy.

Comment (staktrace) [reply gfx/layers/opengl/GLManager.cpp]: Ah, makes sense. Thanks for clarifying

**Comment by mephisto41** [comment]
All dependent bugs have been merged. So I change title of this PR to separate bindings to different crate.

**Pull request closed without merging**